<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [1]:
import pandas as pd
import numpy as np

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [2]:
df = pd.read_csv("degreeClassification.csv")
df.head(20)

Title  \
0                   Location   
1             Academic years   
2                Data source   
3           Data source link   
4   Data file canonical link   
5                    Licence   
6                  Code page   
7                 Disclaimer   
8              2013/14 total   
9              2014/15 total   
10             2015/16 total   
11             2016/17 total   
12             2017/18 total   
13              Last updated   
14                       NaN   
15                       NaN   
16             Academic Year   
17                   2013/14   
18                   2013/14   
19                   2013/14   

   Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider  \
0                                                  UK                                                                     
1                                  2013/14 to 2017/18                                                                     
2                                                HESA                                                                     
3   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
4   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
5   Creative Commons Attribution 4.0 International...                                                                     
6                                      Unicode UTF-8                                                                      
7   Please note that this data includes rounded to...                                                                     
8                                              395605                                                                     
9                                              370905                                                                     
10                                             376330                                                                     
11                                             391875                                                                     
12                                             396940                                                                     
13                                             Jan-19                                                                     
14                                                NaN                                                                     
15                                                NaN                                                                     
16                                      Mode of study                                                                     
17                                                All                                                                     
18                                                All                                                                     
19                                                All                                                                     

   Unnamed: 2              Unnamed: 3                      Unnamed: 4  \
0         NaN                     NaN                             NaN   
1         NaN                     NaN                             NaN   
2         NaN                     NaN                             NaN   
3         NaN                     NaN                             NaN   
4         NaN                     NaN                             NaN   
5         NaN                     NaN                             NaN   
6         NaN                     NaN                             NaN   
7         NaN                     NaN                             NaN   
8         NaN                     NaN                             NaN   
9         NaN                     Na

In [3]:
df.dtypes

Title                                                                                                                   object
Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider    object
Unnamed: 2                                                                                                              object
Unnamed: 3                                                                                                              object
Unnamed: 4                                                                                                              object
Unnamed: 5                                                                                                              object
Unnamed: 6                                                                                                              object
dtype: object

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [4]:
df = df.iloc[17:]
column_names = ['Academic Year', 'Mode of Study', 'Sex', 'Country of HE Provider', 'Classification of First Degree', 'Number', 'Percentage']
df.columns = column_names
df.head(20)

Academic Year Mode of Study  Sex Country of HE Provider  \
17       2013/14           All  All                    All   
18       2013/14           All  All                    All   
19       2013/14           All  All                    All   
20       2013/14           All  All                    All   
21       2013/14           All  All                    All   
22       2013/14           All  All                England   
23       2013/14           All  All                England   
24       2013/14           All  All                England   
25       2013/14           All  All                England   
26       2013/14           All  All                England   
27       2013/14           All  All       Northern Ireland   
28       2013/14           All  All       Northern Ireland   
29       2013/14           All  All       Northern Ireland   
30       2013/14           All  All       Northern Ireland   
31       2013/14           All  All       Northern Ireland   
32       2013/14           All  All               Scotland   
33       2013/14           All  All               Scotland   
34       2013/14           All  All               Scotland   
35       2013/14           All  All               Scotland   
36       2013/14           All  All               Scotland   

   Classification of First Degree  Number Percentage  
17                          First   79410        20%  
18                   Lower second   95965        24%  
19                     Third/Pass   21905         6%  
20               Total classified  395605        NaN  
21                   Upper second  198320        50%  
22                          First   68590        20%  
23                   Lower second   81665        24%  
24                     Third/Pass   19385         6%  
25               Total classified  339170        NaN  
26                   Upper second  169530        50%  
27                          First    1590        19%  
28                   Lower second    2215        26%  
29                     Third/Pass     285         3%  
30               Total classified    8560        NaN  
31                   Upper second    4470        52%  
32                          First    5125        21%  
33                   Lower second    5505        22%  
34                     Third/Pass     875         4%  
35               Total classified   24880        NaN  
36                   Upper second   13380        54%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [5]:
df['Mode of Study'].unique()

array(['All', 'Full-time', 'Part-time'], dtype=object)

<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [6]:
mask = (df['Mode of Study'] == 'All')

<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [7]:
mask = (df['Mode of Study'] == 'All')

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [8]:
df = df[mask]
df['Mode of Study'].unique()

array(['All'], dtype=object)

<ol start='5'>
<li>Output the dataframe</li>

In [9]:
df

Academic Year Mode of Study    Sex Country of HE Provider  \
17         2013/14           All    All                    All   
18         2013/14           All    All                    All   
19         2013/14           All    All                    All   
20         2013/14           All    All                    All   
21         2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1212       2017/18           All  Other                  Wales   
1213       2017/18           All  Other                  Wales   
1214       2017/18           All  Other                  Wales   
1215       2017/18           All  Other                  Wales   
1216       2017/18           All  Other                  Wales   

     Classification of First Degree  Number Percentage  
17                            First   79410        20%  
18                     Lower second   95965        24%  
19                       Third/Pass   21905         6%  
20                 Total classified  395605        NaN  
21                     Upper second  198320        50%  
...                             ...     ...        ...  
1212                          First      10         ..  
1213                   Lower second       0         ..  
1214                     Third/Pass       0         ..  
1215               Total classified      20        NaN  
1216                   Upper second      10         ..  

[475 rows x 7 columns]

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [10]:
mask_sex = (df['Sex'] == 'All')
mask_country = (df['Country of HE Provider'] == 'All')

df = df[mask_sex]

In [11]:
df = df[mask_country]
df

/tmp/ipykernel_16039/2653832462.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df = df[mask_country]


Academic Year Mode of Study  Sex Country of HE Provider  \
17         2013/14           All  All                    All   
18         2013/14           All  All                    All   
19         2013/14           All  All                    All   
20         2013/14           All  All                    All   
21         2013/14           All  All                    All   
292        2014/15           All  All                    All   
293        2014/15           All  All                    All   
294        2014/15           All  All                    All   
295        2014/15           All  All                    All   
296        2014/15           All  All                    All   
567        2015/16           All  All                    All   
568        2015/16           All  All                    All   
569        2015/16           All  All                    All   
570        2015/16           All  All                    All   
571        2015/16           All  All                    All   
842        2016/17           All  All                    All   
843        2016/17           All  All                    All   
844        2016/17           All  All                    All   
845        2016/17           All  All                    All   
846        2016/17           All  All                    All   
1122       2017/18           All  All                    All   
1123       2017/18           All  All                    All   
1124       2017/18           All  All                    All   
1125       2017/18           All  All                    All   
1126       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
17                            First   79410        20%  
18                     Lower second   95965        24%  
19                       Third/Pass   21905         6%  
20                 Total classified  395605        NaN  
21                     Upper second  198320        50%  
292                           First   81640        22%  
293                    Lower second   85250        23%  
294                      Third/Pass   20340         5%  
295                Total classified  370905        NaN  
296                    Upper second  183680        50%  
567                           First   88890        24%  
568                    Lower second   81595        22%  
569                      Third/Pass   19280         5%  
570                Total classified  376330        NaN  
571                    Upper second  186565        50%  
842                           First  100945        26%  
843                    Lower second   79515        20%  
844                      Third/Pass   19025         5%  
845                Total classified  391875        NaN  
846                    Upper second  192395        49%  
1122                          First  110475        28%  
1123                   Lower second   76210        19%  
1124                     Third/Pass   17830         4%  
1125               Total classified  396940        NaN  
1126                   Upper second  192425        48%

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [12]:
mask_class = df['Classification of First Degree'] != 'Total classified'
df = df[mask_class]
df

Academic Year Mode of Study  Sex Country of HE Provider  \
17         2013/14           All  All                    All   
18         2013/14           All  All                    All   
19         2013/14           All  All                    All   
21         2013/14           All  All                    All   
292        2014/15           All  All                    All   
293        2014/15           All  All                    All   
294        2014/15           All  All                    All   
296        2014/15           All  All                    All   
567        2015/16           All  All                    All   
568        2015/16           All  All                    All   
569        2015/16           All  All                    All   
571        2015/16           All  All                    All   
842        2016/17           All  All                    All   
843        2016/17           All  All                    All   
844        2016/17           All  All                    All   
846        2016/17           All  All                    All   
1122       2017/18           All  All                    All   
1123       2017/18           All  All                    All   
1124       2017/18           All  All                    All   
1126       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
17                            First   79410        20%  
18                     Lower second   95965        24%  
19                       Third/Pass   21905         6%  
21                     Upper second  198320        50%  
292                           First   81640        22%  
293                    Lower second   85250        23%  
294                      Third/Pass   20340         5%  
296                    Upper second  183680        50%  
567                           First   88890        24%  
568                    Lower second   81595        22%  
569                      Third/Pass   19280         5%  
571                    Upper second  186565        50%  
842                           First  100945        26%  
843                    Lower second   79515        20%  
844                      Third/Pass   19025         5%  
846                    Upper second  192395        49%  
1122                          First  110475        28%  
1123                   Lower second   76210        19%  
1124                     Third/Pass   17830         4%  
1126                   Upper second  192425        48%

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

In [13]:
df.drop(columns=['Mode of Study', 'Sex', 'Country of HE Provider'], inplace = True)
df

/tmp/ipykernel_16039/2230704610.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=['Mode of Study', 'Sex', 'Country of HE Provider'], inplace = True)


Academic Year Classification of First Degree  Number Percentage
17         2013/14                          First   79410        20%
18         2013/14                   Lower second   95965        24%
19         2013/14                     Third/Pass   21905         6%
21         2013/14                   Upper second  198320        50%
292        2014/15                          First   81640        22%
293        2014/15                   Lower second   85250        23%
294        2014/15                     Third/Pass   20340         5%
296        2014/15                   Upper second  183680        50%
567        2015/16                          First   88890        24%
568        2015/16                   Lower second   81595        22%
569        2015/16                     Third/Pass   19280         5%
571        2015/16                   Upper second  186565        50%
842        2016/17                          First  100945        26%
843        2016/17                   Lower second   79515        20%
844        2016/17                     Third/Pass   19025         5%
846        2016/17                   Upper second  192395        49%
1122       2017/18                          First  110475        28%
1123       2017/18                   Lower second   76210        19%
1124       2017/18                     Third/Pass   17830         4%
1126       2017/18                   Upper second  192425        48%

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

In [14]:
df.dtypes

Academic Year                     object
Classification of First Degree    object
Number                            object
Percentage                        object
dtype: object

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

In [15]:
for i in range(len(df)) :
    df.iloc[i]['Percentage'] = df.iloc[i]['Percentage'][:-1]

df

Academic Year Classification of First Degree  Number Percentage
17         2013/14                          First   79410         20
18         2013/14                   Lower second   95965         24
19         2013/14                     Third/Pass   21905          6
21         2013/14                   Upper second  198320         50
292        2014/15                          First   81640         22
293        2014/15                   Lower second   85250         23
294        2014/15                     Third/Pass   20340          5
296        2014/15                   Upper second  183680         50
567        2015/16                          First   88890         24
568        2015/16                   Lower second   81595         22
569        2015/16                     Third/Pass   19280          5
571        2015/16                   Upper second  186565         50
842        2016/17                          First  100945         26
843        2016/17                   Lower second   79515         20
844        2016/17                     Third/Pass   19025          5
846        2016/17                   Upper second  192395         49
1122       2017/18                          First  110475         28
1123       2017/18                   Lower second   76210         19
1124       2017/18                     Third/Pass   17830          4
1126       2017/18                   Upper second  192425         48

In [16]:
df['Percentage'] = df['Percentage'].astype(int)
df['Number'] = df['Number'].astype(int)
df['Academic Year'] = df['Academic Year'].astype(str)
df['Classification of First Degree'] = df['Classification of First Degree'].astype(str)

df.dtypes

/tmp/ipykernel_16039/3166726546.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Percentage'] = df['Percentage'].astype(int)
/tmp/ipykernel_16039/3166726546.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Number'] = df['Number'].astype(int)
/tmp/ipykernel_16039/3166726546.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/panda

Academic Year                     object
Classification of First Degree    object
Number                             int64
Percentage                         int64
dtype: object

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

In [17]:
mask_first = df['Classification of First Degree'] == 'First'
mask_upsec = df['Classification of First Degree'] == 'Upper second'
mask_lowsec = df['Classification of First Degree'] == 'Lower second'
mask_third = df['Classification of First Degree'] == 'Third/Pass'

df_first = df[mask_first]
df_upp_sec = df[mask_upsec]
df_low_sec = df[mask_lowsec]
df_third = df[mask_third]

<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.

In [18]:
df_first.to_csv('df_first.csv')
df_upp_sec.to_csv('df_upp_sec.csv')
df_low_sec.to_csv('df_low_sec.csv')
df_third.to_csv('df_third.csv')
df.to_csv('df.csv')